util.go

```diff
@@ -38,6 +37,28 @@ func min(values ...uint32) uint32 {
 	return m
 }
 
+// The segmented buffer looks like:
+//
+//      |     data      | empty space       |
+//       < window (10)                     >
+//       < len (5)     > < cap (5)         >
+//                       < pending (4) ->
+//
+// As data is read, the buffer gets updated like so:
+//
+//         |     data   | empty space       |
+//          < window (8)                   >
+//          < len (3)  > < cap (5)         >
+//                       < pending (4) ->
+//
+// It can then grow as follows (given a "max" of 10):
+//
+//
+//         |     data   | empty space          |
+//          < window (10)                     >
+//          < len (3)  > < cap (7)            >
+//                       < pending (4) ->
```

Comment:
This comment block is super useful for understanding :)
Can you add a note about how `pending` relates to the other numbers?
e.g - new data coming in will reduce pending as it moves more data from cap to length, right?

Reply:
Ah, yes, you're right.